### PR TITLE
UCP/CORE: Reduce size of generic EP extension for further usage

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Reduce the size of generic EP extension for further usage

## Why ?

Generic EP extension will be used by #5821 to have `ucs_hlist_head_t` for tracking UCP requests.

## How ?

Introduce a pointer to a control extension that is placed in the generic extension. Control extension includes:
- local and remote IDs
- error callback
- union for `close_req` and `listener`
other fields are in generic extension, since they are could not be moved to a control extension (e.g. we have to have a mechanism to translate `flush_state` or `conn_match` to generic extension, or even to EP)

### CODE REVIEW

**Comment by dmitrygx** [comment]
@hoopoepg @brminich @evgeny-leksikov could you review pls?

**Review by evgeny-leksikov — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -99,28 +99,29 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
...
-        ucs_error("Failed to allocate ep keys");
+    ep->cfg_index                      = UCP_WORKER_CFG_INDEX_NULL;
+    ep->worker                         = worker;
+    ep->am_lane                        = UCP_NULL_LANE;
+    ep->flags                          = 0;
+    ep->conn_sn                        = UCP_EP_MATCH_CONN_SN_MAX;
+    ucp_ep_ext_gen(ep)->user_data      = NULL;
+    ucp_ep_ext_gen(ep)->control_ext    = ucs_calloc(
+                                             1, sizeof(ucp_ep_ext_control_t),
+                                             "ep_control_ext");
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:110]: indentation

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -99,28 +99,29 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
...
-                                               "ep_ids");
-    if (ucp_ep_ext_gen(ep)->ids == NULL) {
-        ucs_error("Failed to allocate ep keys");
+    ep->cfg_index                      = UCP_WORKER_CFG_INDEX_NULL;
+    ep->worker                         = worker;
+    ep->am_lane                        = UCP_NULL_LANE;
+    ep->flags                          = 0;
+    ep->conn_sn                        = UCP_EP_MATCH_CONN_SN_MAX;
+    ucp_ep_ext_gen(ep)->user_data      = NULL;
+    ucp_ep_ext_gen(ep)->control_ext    = ucs_calloc(
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:108]: can reduce number of spaces

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -99,28 +99,29 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
...
-    ucp_ep_ext_gen(ep)->ids->remote = UCP_EP_ID_INVALID;
+    ucp_ep_ext_control(ep)->err_cb     = NULL;
+    ucp_ep_ext_control(ep)->ids.local  = UCP_EP_ID_INVALID;
+    ucp_ep_ext_control(ep)->ids.remote = UCP_EP_ID_INVALID;
 
-    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen(ep)->ep_match) >=
-                      sizeof(ucp_ep_ext_gen(ep)->listener));
     UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen(ep)->ep_match) >=
                       sizeof(ucp_ep_ext_gen(ep)->flush_state));
+    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_control(ep)->listener) ==
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:123]: why is it required?

File: src/ucp/core/ucp_ep.h [hunk 4]
```diff
@@ -393,31 +393,40 @@ typedef struct {
...
-                                                    stride allocator improvement */
-    void                          *user_data;    /* User data associated with ep */
-    ucs_list_link_t               ep_list;       /* List entry in worker's all eps list */
+    ucp_ep_ids_t                  ids;           /* Local and remote IDs */
     ucp_err_handler_cb_t          err_cb;        /* Error handler */
 
-    /* Endpoint match context and remote completion status are mutually exclusive,
-     * since remote completions are counted only after the endpoint is already
-     * matched to a remote peer.
+    /* Endpoint match context is mutually exclusive with:
```
Comment (brminich) [inline src/ucp/core/ucp_ep.h:403]: is this comment relevant now? match is not here and they are not in the union anymore

File: src/ucp/core/ucp_ep.inl [hunk 5]
```diff
@@ -216,10 +222,10 @@ static inline void ucp_ep_flush_state_reset(ucp_ep_h ep)
                 (flush_state->cmpl_sn == 0) &&
                 ucs_queue_is_empty(&flush_state->reqs)));
 
+    ucs_queue_head_init(&flush_state->reqs);
     flush_state->send_sn = 0;
     flush_state->cmpl_sn = 0;
-    ucs_queue_head_init(&flush_state->reqs);
-    ep->flags |= UCP_EP_FLAG_FLUSH_STATE_VALID;
+    ep->flags           |= UCP_EP_FLAG_FLUSH_STATE_VALID;
```
Comment (brminich) [inline src/ucp/core/ucp_ep.inl:228]: unrelated?

File: src/ucp/core/ucp_ep.c (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c]: done

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c]: fixed

File: src/ucp/core/ucp_ep.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c]: it's not needed
removed

File: src/ucp/core/ucp_ep.inl (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.inl]: removed

File: src/ucp/core/ucp_ep.h (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h]: yes, should remove it
and use the old one

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Review by hoopoepg — APPROVED** [review]

**Comment by dmitrygx** [comment]
the failure is #5856

**Comment by dmitrygx** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_ep.h [hunk 6]
```diff
@@ -393,31 +393,38 @@ typedef struct {
 } ucp_ep_ids_t;
 
 
-/*
+/**
+ * Endpoint extension for control data path
+ */
+typedef struct {
+    ucp_ep_ids_t                  ids;           /* Local and remote IDs */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:400]: extra pointer deref for rndv/am protocols?

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:400]: I see that remote_id is used for AMO/RMA/Stream. Is it used for rndv/am as well?
looks like we have to have it in `ucp_ep_ext_gen_t`

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:400]: btw, it was a double pointer deref before this PR, since IDs were allocated on a heap
`ucp_ep_ext_gen_t*` -> `ucp_ep_ids_t*`
with this PR we have:
`ucp_ep_ext_gen_t*` -> `ucp_ep_ext_control_t*` and offset to the [0] or [1] ID

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:400]: but if we would like to avoid it
we need to allocate 3 strides instead of having a pointer to control extension in general extension + we could move `ucp_ep_match_elem_t` field there (it is on general extension now, since we have to have an access to EP in UCP/EP_MATCH code)
wdyt?

Comment (yosefe) [reply src/ucp/core/ucp_ep.h:400]: can we move ep id's to a stride extension and all the rest to the heap?

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h]: decoupled local and remote IDs as dicussed

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@yosefe could you reivew pls?

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe do we need to restart IO demo and RTD somehow?

**Comment by yosefe** [comment]
> @yosefe do we need to restart IO demo and RTD somehow?

no need

**Comment by dmitrygx** [comment]
failure is #5875 
bot:pipe:retest

**Comment by dmitrygx** [comment]
IO demo and RTD failures are unrelated
@yosefe ok to merge?
